[aot] change method table generation so that it works for larger assemblies on ARM

lewurm opened the pull request:
At the end of the AOT compilation, we dump a table containing all method
entrypoints using the toolchain linker to resolve the addresses. Previously
we used `bl` for that, however on larger assemblies the offset might
overflow the 24bit immediate (thus, the linker bailing out with
"Error: branch out of range"). Instead, using `ldr pc,=<label>` followed
by `.ltorg` emits the full 32bit address to a method. That takes 8 bytes
for each method instead of 4 bytes.

cc @vargaz @ludovic-henry

vargaz commented:
The reason we emit a bl is that all platforms can handle that relocation correctly, we tried various other approaches, and there were problems on ios. etc. The linker is supposed to insert thunks to resolve the problem, couldn't the linux linker forced to do this ?

lewurm commented:
I guess you're talking about "branch islands", right? Looks like `ld64` supports that and seems to be enable by default.

I don't find a similar feature for binutils :/

vargaz commented:
I'm okay with this if its inside an #ifdef TARGET_ANDROID or something.

lewurm commented:
rebased and added android ifdef.  Hey @ludovic-henry, was that something you hit again in your hybrid aot work? do you think we should get this into c9?

ludovic-henry commented:
@lewurm I don't remember hitting that on hybrid aot (or that could be the case I don't understand exactly what your change does)
